### PR TITLE
POC implementation of extension loading in Selenium

### PR DESCRIPTION
**Pull request by fregante**
Based on #711 

I added some code to be able to run the tests locally (if the Browserstack keys are missing). That works, the extension loads ✅ 

I also followed the instructions to upload the extension to browserstack and I think this is not possible due to memory limits: ❌

```
(node:2541) UnhandledPromiseRejectionWarning: WebDriverError: Could not start Browser / Emulator
 Reason: 
Java heap space
    at Object.checkLegacyResponse (node_modules/selenium-webdriver/lib/error.js:553:15)
    at parseHttpResponse (node_modules/selenium-webdriver/lib/http.js:634:13)
    at Executor.execute (node_modules/selenium-webdriver/lib/http.js:568:28)
```

This is because the 19MB zipped _production_ extension needs to be base64’d which turns it into a 30MB string that needs to be uploaded when the tests launch… once per environment. 

I double-checked by uploading Refined GitHub and it was loaded onto Browserstack’s browser correctly.

_Side note: The upload of this file takes 4 minutes for me even though it should take 15 seconds._

Thoughts?

## Local tests

```js
npm i 
npm run build # or watch
node selenium/sample-build.js
```

## Browserstack tests

```
npm run build
BROWSERSTACK_USERNAME=XXX BROWSERSTACK_ACCESS_KEY=XXX node selenium/sample-build.js
```

### CODE REVIEW

File: selenium/sample-build.js [hunk 1]
```diff
@@ -32,61 +39,60 @@ const accessKey = process.env.BROWSERSTACK_ACCESS_KEY;
...
+    ...(capabilities["browser"] && { browserName: capabilities["browser"] }), // Because NodeJS language binding requires browserName to be defined
+  });
+
+  const driver = await builder.build();
   await driver.get("https://www.pixiebrix.com");
   try {
     await driver.wait(webdriver.until.titleMatches(/pixiebrix/i), 5000);
     console.log(await driver.getTitle());
-    await driver.executeScript(
-      'browserstack_executor: {"action": "setSessionStatus", "arguments": {"status":"passed","reason": "Title contains PixieBrix!"}}'
```
Comment (fregante) [inline selenium/sample-build.js:51]: These commands are BS-specific so they don't run locally. Temporarily disabled.

File: selenium/sample-build.js [hunk 2]
```diff
@@ -15,11 +15,18 @@
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
  */
 
+const fs = require("fs");
 const webdriver = require("selenium-webdriver");
+const chrome = require("selenium-webdriver/chrome");
```
Comment (twschiller) [inline selenium/sample-build.js:20]: Codacy has a fix for the issue: [Do not use "require".](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7734019)
```suggestion
import chrome from "selenium-webdriver/chrome";
```

File: selenium/sample-build.js [hunk 3]
```diff
@@ -32,61 +39,60 @@
...
+    };
+  }
+
+  void testPixieBrixHomepage(capabilities1);
+}
 
-void testPixieBrixHomepage(capabilities1);
-void testPixieBrixHomepage(capabilities2);
-void testPixieBrixHomepage(capabilities3);
+init();
```
Comment (twschiller) [inline selenium/sample-build.js:98]: Codacy has a fix for the issue: [Promises must be handled appropriately or explicitly marked as ignored with the `void` operator.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7734019)
```suggestion
void init();
```

File: selenium/sample-build.js [hunk 4]
```diff
@@ -15,11 +15,18 @@
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
  */
 
+const fs = require("fs");
```
Comment (twschiller) [inline selenium/sample-build.js:18]: Codacy has a fix for the issue: [Do not use "require".](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7734019)
```suggestion
import fs from "fs";
```

**Comment by twschiller** [comment]
@fregante Thanks for looking into this - I've submitted a help request with BrowserStack to see if there's any information they can provide

FWIW, the artifact that gets uploaded to GitHub is only 13MB. Maybe you could squeeze of some of the size by tweaking the zip parameters? I doubt that will fix the issue though

I still think its worth while getting the infrastructure working locally though in the meantime on both Chrome and FF.

**Comment by twschiller** [comment]
@fregante Try again with BrowserStack when you have a chance. Their support team tweaked a setting on their end. (I also emailed you their response)

**Comment by fregante** [comment]
I was able to upload the 18MB zipfile locally but the same Heap problem seems to be present when run on GHA.

Also can you remind me why we're routing this via browserstack? The screen recording tool seems nice but this is adding an extra layer of indirection and slowing down any testing and test-writing. What do you think if we start without BS now and then reconsider it later?

File: src/frameworks/contrib/ember.ts [hunk 5]
```diff
@@ -149,22 +146,20 @@
...
     if (isMutableCell(value) && "value" in value) {
       return traverse(value.value);
     } else if ("_cache" in value) {
       return traverse(value._cache);
     } else if (Array.isArray(value.content)) {
       // consider arrays a traverse because knowing the property name by itself isn't useful for anything
-      return value.content.map(traverse);
-    } else {
-      return mapValues(pickExternalProps(value), recurse);
+      return value.content.map((x: any) => traverse(x));
```
Comment (twschiller) [inline src/frameworks/contrib/ember.ts:157]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/pixiebrix/pixiebrix-extension/pullRequest?prid=7734019)
```suggestion
      return value.content.map((x: unknown) => traverse(x));
```

**Comment by fregante** [comment]
I cleaned up the file a bit to make senses of the many options available across browsers and to be able to run tests locally or on browserstack, for now. There are many ways to do things so I had to look into the source of a few projects to determine how things work and what the API hides.

## Firefox

I added support for Firefox **but** strangely it's not able to load the zipped extension because GeckoDriver throws an error:

```
(node:32024) UnhandledPromiseRejectionWarning: WebDriverError: Could not install add-on: /var/folders/jx/zxcvz4sj195b3vf9l_7ycjsm0000gn/T/addon-01614d25-c97d-45d8-a070-48c1a5ea0e0d.xpi: ERROR_CORRUPT_FILE: The file appears to be corrupt.
    at Object.throwDecodedError (./node_modules/selenium-webdriver/lib/error.js:517:15)
    at parseHttpResponse (./node_modules/selenium-webdriver/lib/http.js:642:13)
    at Executor.execute (./node_modules/selenium-webdriver/lib/http.js:568:28)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Driver.execute (./node_modules/selenium-webdriver/lib/webdriver.js:731:17)
```

I'm feeding it a zip file and it copies it to another temp location renaming it to xpi. I'll try to figure out whether it expects an xpi exclusively, but I'm able to load the zip manually so I'm not sure what the error is.

## Documentation

Here are some documentation pages that I found along the way. I'll sort them out once we figure out what's useful.


BrowserStack-specific Selenium capabilities: https://www.browserstack.com/automate/capabilities

WebDriver Selenium capabilities: https://github.com/SeleniumHQ/selenium/wiki/DesiredCapabilities

ChromeOptions: https://sites.google.com/a/chromium.org/chromedriver/capabilities

Firefox Options: https://firefox-source-docs.mozilla.org/testing/geckodriver/Capabilities.html#capabilities-example

GeckoDriver docs: https://www.selenium.dev/selenium/docs/api/javascript/module/selenium-webdriver/firefox.html

ChromeDriver docs https://www.selenium.dev/selenium/docs/api/javascript/module/selenium-webdriver/chrome.html

**Comment by twschiller** [comment]
Thanks for the references, I created a wiki page we can use to collect our notes: https://github.com/pixiebrix/pixiebrix-extension/wiki/Web-Extension-Testing-Reference

> Also can you remind me why we're routing this via browserstack?

BrowserStack has a couple nice features: 1) running multiple versions of the browsers, 2) you can watch the video recordings

> What do you think if we start without BS now and then reconsider it later?

I would be OK with that if BrowserStack is a barrier

> not able to load the zipped extension

Hmm, that's odd the manual loading of the zip works. Perhaps the webdriver has limitations on the ZIP format (headers, encoding, etc.). I'll check things out and look at the code to see if I notice anything

**Comment by fregante** [comment]
> What do you think of using TypeScript for writing the file?

Might be good, but it might also be a pain if it doesn't work well (that is, if types are incomplete or difficult, especially with mocks). We can give it a try.



> 1) running multiple versions of the browsers

Same for the regular web drivers, except the tests might take a bit longer because multiple version need to be downloaded (which might actually be as long as it takes to upload the extension anyway)

> 2) you can watch the video recordings

Those are useful but you can get the same result by running it on you computer while also debugging the issue that appears.

They're all good as long as we manage to actually run the tests efficiently without spending too much time on fixing Browserstack itself.



> Looks like I might need to sign the add-on: [mozilla/geckodriver#473](https://github.com/mozilla/geckodriver/issues/473)?

Note: That issue is pre-webextensions on Firefox, which were released later in 2017. I will try with a signed extension anyway. The weird part is that it does work with regular Firefox locally via WebDriver, but not via BS.

> When I try to run on Firefox (using just the output of `npm run watch` or also `npm run build`, I'm getting an ERROR_SIGNEDSTATE_REQUIRED error. Am I getting farther than you?

Strange… I'm seeing that error locally now as well 🤔 that was not the case before.

**Comment by fregante** [comment]
I've been researching this on and off for a while now, hoping to wake up one day and figure it out. 😅 Here's what I learned:

- both ZIP and XPI are compatible
- indeed Firefox requires the add-on to be signed: I downloaded the signed Refined GitHub and it worked; I removed the META-INF folder and I got the error
- there's "no way" to install unsigned addons in regular Firefox, `xpinstall.signatures.required` is ignored
- only Firefox Developer Edition and Nightly allow unsigned addons
- Browserstack does not have either one of those
- Signing requires setting up AMO API keys locally and using a different extension version every time, I think
- It's done by uploading the app via `web-ext sign`, which means 12MB need to be uploaded to 2 services before starting the tests

In the end I found this old repo that has some GeckoDriver-specific code that enables the loading of unsigned extensions that we might still be able to adjust to our use:

https://github.com/mozilla/example-addon-repo/blob/b45f244263252633b05282e1275004fc6ede5ffa/test/functional/utils.js#L66-L92

`web-ext` is also somehow able to load unsigned extensions but I wasn't immediately able to locate the magic that allows that. It seems that they use a different API from WebDriver so we probably can't use it on browserstack. `web-ext run` was actually Puppeteer’s suggested setup as well.

Either way that repo is a good starting point for _Selenium mocha WebDriver tests in Firefox,_ but I might want to look for something a bit more recent than 2016 if that also fails.

That's all the updates I have for now, I just thought I'd let you know.

**Pull request closed without merging**